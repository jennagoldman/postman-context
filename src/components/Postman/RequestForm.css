.formContainer {
  margin: 1rem 0;
  width: 55rem;
  display: inline-block;
}

.formSection {
  width: 55rem;
}

.urlDiv {
  display: inline-block;
  width: 35rem;
}

.urlInput {
  width: 32rem;
  padding: 0.25rem;
  margin-bottom: 1rem;
  display: inline-block;
}

.methodsDiv {
  display: inline-block;
  width: 20rem;
}

fieldset {
  border: 1px solid #fc40e6;
}

.methodsDiv fieldset {
  padding: 1.3rem 1rem 2rem 1rem;
}

input {
  margin: 0.75rem;
  border: 1px solid #fc40e6;
  font-size: 1em;
  font-family: 'Cairo', sans-serif;
}

input[type="radio"] {
  display: none;
}

input:checked + label {
  background: #fc40e6;
  border: 1px solid #fc40e6;
  color: #fff;
  font-weight: 600;
}

input + label {
  text-transform: uppercase;
  padding:  0.1rem 0.5rem;
  margin: 0.35rem;
  cursor: pointer;
  border: 1px solid #fc40e6;
  font-size: 1em;
}

textarea {
  width: 51rem;
  height:15rem;
  margin: 1rem;
  padding: 1rem;
  border: 1px solid #fc40e6;
  font-size: 1em;
  font-family: 'Cairo', sans-serif;
  line-height: 1.25rem;
}

@media only screen and (max-width: 1023px) {
  .urlDiv, .methodsDiv {
    display: block;
    width: 55rem;
  }

}