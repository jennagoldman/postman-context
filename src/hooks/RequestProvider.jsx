import React, { createContext, useState, useEffect, useContext } from 'react';
import makeRequest from '../services/make-request.js';
import PropTypes from 'prop-types';
import RequestForm from '../components/Postman/RequestForm.jsx';

export const RequestContext = createContext();

export const RequestProvider = ({ children }) => {
  const [url, setUrl] = useState('');
  const [method, setMethod] = useState('');
  const [body, setBody] = useState('');
  const [response, setResponse] = useState('');
  const [history, setHistory] = useState([]);

  const handleChange = ({ target }) => {
    if(target.name === 'url') setUrl(target.value);
    if(target.name === 'method') setMethod(target.value);
    if(target.name === 'body') setBody(target.value);
  }

  useEffect(() => {
    if(localStorage.getItem('requestHistory') === null) {
      localStorage.setItem('requestHistory', JSON.stringify([]));
    }
    const history = localStorage.getItem('requestHistory');
    setHistory(JSON.parse(history));
  }, []);

  useEffect(() => {
    localStorage.setItem('requestHistory', JSON.stringify(history));
  }, [history]);

  const handleRequestSubmit = (event) => {
    event.preventDefault();
    makeRequest(url, method, body)
      .then(response => setResponse(response));

    setHistory(prevHistory => ([
      ...prevHistory,
      {
        url,
        method,
        body
      }
    ]));
    localStorage.setItem('requestHistory', JSON.stringify(history));
  };

  const handleHistoryReset = () => {
    localStorage.setItem('requestHistory', JSON.stringify([]));
    setHistory([]);
  };

  return (
    <RequestContext.Provider value={{ url, method, body, response, history, handleChange, handleRequestSubmit, handleHistoryReset}}>
      {children}
    </RequestContext.Provider>
  )
}

RequestProvider.propTypes = {
  children: PropTypes.node
}

export const useRequestContext = () => {
  const context = useContext(RequestContext);
  return context;
}

export const useResponse = () => {
  const { response } = useContext(RequestContext);
  return response;
}