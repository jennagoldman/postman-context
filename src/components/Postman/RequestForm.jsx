import React from 'react';
import { useRequestContext } from '../../hooks/RequestProvider.jsx';

const RequestForm = () => {
  const { url, method, body, handleChange, handleRequestSubmit } = useRequestContext();

  return (
  <section className={styles.formContainer}>
    <form onSubmit={handleRequestSubmit}>
      <section className={styles.formSection}>
        <div className={styles.urlDiv}>
          <fieldset>
            <legend>URL</legend>
            <input 
              className={styles.urlInput} 
              type="text" 
              name="url" 
              value={url} 
              id="url" 
              onChange={handleChange} 
            />
          </fieldset>
        </div>
        <div className={styles.methodsDiv}>
          <fieldset>
            <legend>METHOD</legend>
            <input 
              type="radio" 
              name="method" 
              value="get" 
              id="get" 
              onChange={handleChange} 
              checked={method === 'get'} 
            />
            <label htmlFor="get">Get</label>
            <input 
              type="radio" 
              name="method" 
              value="put" 
              id="put" 
              onChange={handleChange} 
              checked={method === 'put'}
            />
            <label htmlFor="put">Put</label>
            <input 
              type="radio" 
              name="method" 
              value="post" 
              id="post" 
              onChange={handleChange} 
              checked={method === 'post'} 
            />
            <label htmlFor="post">Post</label>
            <input 
              type="radio" 
              name="method" 
              value="delete" 
              id="delete" 
              onChange={handleChange} 
              checked={method === 'delete'} 
            />
            <label htmlFor="delete">Delete</label>
          </fieldset>
        </div>
      </section>
      <section>
        <fieldset>
          <legend>REQUEST BODY</legend>
          <textarea 
            name="body"
            value={body} 
            onChange={handleChange} 
            placeholder="Raw JSON" />
        </fieldset>
      </section>
    
      <button>Submit</button>
    </form>
  </section>
);
};