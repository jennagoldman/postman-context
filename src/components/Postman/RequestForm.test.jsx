  
import React from 'react';
import { shallow } from 'enzyme';
import { RequestProvider } from '../../hooks/RequestProvider.jsx';
import RequestForm from './RequestForm.jsx';

describe('Header', () => {
  it('matches a snapshot', () => {
    const wrapper = shallow(
      <RequestProvider>
        <RequestForm />
      </RequestProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});