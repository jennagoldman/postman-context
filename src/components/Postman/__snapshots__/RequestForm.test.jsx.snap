// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header matches a snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "body": "",
      "handleChange": [Function],
      "handleHistoryReset": [Function],
      "handleRequestSubmit": [Function],
      "history": Array [],
      "method": "",
      "response": "",
      "url": "",
    }
  }
>
  <RequestForm />
</ContextProvider>
`;
